### DISCUSSION TITLE
Disallow `ref` arguments on readonly instance members

### DISCUSSION THREAD

**Discussion by MithrilMan** [dotnet/csharplang#5042] [Language Ideas]
In _readonly instance members_ motivation section of the specification <https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/readonly-instance-members#motivation> it's stated that the features is done to prevent a copy when using the struct as an `in` parameter

The code below, defeats the modifier purpose:

```
using System;
					
public class Program
{
	public static void Main()
	{
		Point a = new Point();
		a.FakeReadOnly(ref a);
	}
}

public struct Point
{
    public double X { get; set; }
    public double Y { get; set; }
	
	public readonly double FakeReadOnly(ref Point p) {
		p.X= this.X + 5;
		return p.X;
	}
}
```

Allowing this kind of code like may causes unintended side effects if the method is called when you pass the struct as `in` because the "readonly protection" doesn't protect this case, so I think it's worth at least a warning or a mention in the doc.

IMO I'd disallow the use of ref, on a readonly instance method, when the ref argument has the same type of the Type declaring the readonly method.

I can't see why in this use case you should pass a `ref` on readonly methods instead of `in`.

**Comment by khm1600** [comment]
> Allowing this kind of code like may causes unintended side effects if the method is called when you pass the struct as in

How? You cannot pass a `in` into a `ref`.

**Reply by MithrilMan** [reply]
I was talking about passing the struct with that hacked readonly method as an `in` parameter of another method as per specification motivation

**Reply by khm1600** [reply]
You cannot:
```cs
void InParam(in Point p)
{
    p.FakeReadOnly(ref p); // error CS8329: Cannot use variable 'in Point' as a ref or out value because it is a readonly variable
}
```

**Reply by MithrilMan** [reply]
good point, doesn't this mean that a readonly method with a ref type of the same Type cannot be part of the motivation of the feature and so doesn't have a reason to exist?

**Comment by svick** [comment]
Previous discussion about this: https://github.com/dotnet/docs/issues/25622.

I think there are two problems with this proposal:

1. It would be an unacceptable breaking change. C# tries hard to ensure old code continues to compile with newer compilers and breaking that requires a really good justification. "It's a bit confusing", which is effectively the justification here is not good enough. Though if it was just a warning, not an error, that would be much easier to accept.
2. It doesn't actually ensure that the struct does not change. Consider this code:

    ```c#
    using System;
					
    public class Program
    {
    	public static void Main()
    	{
    		Point.P.FakeReadOnly();
    	}
    }

    public struct Point
    {
    	public static Point P;
    
    	public double X { get; set; }
    	public double Y { get; set; }
	
    	public readonly double FakeReadOnly() {
    		P.X = this.X + 5;
    		return P.X;
    	}
    }
    ```

    This still modifies `this` while a `readonly` method executes, even though it doesn't have any `ref` parameters.

**Reply by MithrilMan** [reply]
I'd like to know if this kind of behavior was considered when the feature was released.
Also having a ref argument of the same time passed to a readonly method smells already as bad design

Having an analyzer rule for that would be already valuable.

your example show another kind of bad behavior (and like my example shows signs of bad design) that's unluckily harder to fix than just "prevent to use ref and force to use in" error/warning because to fix that kind of problem C# would need Ownership

**Reply by jnm2** [reply]
`readonly` has never meant "pure." It has only thus far referred to the mutability of memory regions covered by fields of the declaring type.
